_posts/2023-06-13-http4s-authentication-part2.md

```diff
@@ -0,0 +1,894 @@
+
+#### 3.1.3.0 The EmailService
+
+In this section, we'll create the `EmailService` class to handle sending our QR BarCode to the user.
+
+```scala
+package com.rockthejvm
+
+import com.sendgrid.helpers.mail.objects.Email
+import java.nio.file.Paths
+
+class EmailService(sendto: String):
+    val from = new Email("hkateu@gmail.com")
+    val to = new Email(sendto)
+
+    val subject = "Sending with SendGrid is Fun"
+    val message = s"""
+        |<H3>Website Login</H3>
+        |<p>Please complete the login process using Google Authenticator</p>
+        |<p>Scan the QR Barcode attached and send us the token number.</p>
+        """.stripMargin
+
+    val file = Paths.get("barCode.png")
+    val content = new Content("text/html", message)
+end EmailService
+```
+
+We start by defining some variables needed to send an email. The `from` and `to` variables represent the person sending the email and the person to whom the email will be sent. SendGrid provides an `Email` class to structure these email addresses. Notice that we pass the email address of the person we are sending to through the `EmailService` constructor.
+
+We also define `subject` as a string and `message` as a string containing HTML. The `file` variable is a `Path` to the BarCode png image. `content` is an instance of the `Content` class that tells SendGrid what type of message we are sending, in our case `text/html`.
+
+```scala
+import com.sendgrid.helpers.mail.Mail
+import java.util.Base64
+
+class EmailService(sendto: String):
+    ...
+
+    val mail = new Mail(from, subject, to, content)
+    val attachments = new Attachments()
+    attachments.setFilename(file.getFileName().toString())
+    attachments.setType("application/png")
+    attachments.setDisposition("attachment")
+    val attachmentContentBytes = Files.readAllBytes(file)
+    val attachmentContent = Base64.getEncoder().encodeToString(attachmentContentBytes)
+    attachments.setContent(attachmentContent)
+    mail.addAttachments(attachments)
+end EmailService
+```
+
+SendGrid provides a `Mail` class that takes our `from`, `subject`, `to`, and `content` variables as parameters to structure our email. We then create `attachments` which is an instance of the `Attachments` class, it provides `setFilename()`, `setType()`, and `setDisposition()` methods to structure our attachment.
+
+We use the `readAllBytes()` method from `Files` to convert our png image to type `Array[Byte]` and store it in `attachmentContentBytes`, then encode it to Base64 and store it in `attachmentContent` since SendGrid requires Base64 encoding to send attachments.
+
+We then call `attachments.setContent(attachmentContent)` and `mail.addAttachments(attachments)` to add our file as an attachment to our `Mail` instance.
+
+```scala
+import com.sendgrid.{SendGrid, Method, Request, Response}
+import scala.util.*
+import org.http4s.dsl.io.*
+import cats.effect.IO
+
+class EmailService(sendto: String):
```

Comment:
I think this class can be replaced with a function `send2FA(to: String)`, because it's only used once.
Moreover, this function can return an IO, because the server works with IOs.

Reply:
Added method to OtpInteractiveService and removed EmailService